src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionMappingService.cs

```diff
@@ -0,0 +1,341 @@
+            {
+                SourceDataSetVersionId = liveVersion.Id,
+                TargetDataSetVersionId = nextDataSetVersionId,
+                LocationMappingPlan = locationMappings,
+                FilterMappingPlan = filterMappings
+            });
+
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+        return Unit.Instance;
+    }
+
+    public async Task ApplyAutoMappings(
+        Guid nextDataSetVersionId,
+        CancellationToken cancellationToken = default)
+    {
+        var mappings = await publicDataDbContext
+            .DataSetVersionMappings
+            .SingleAsync(
+                mapping => mapping.TargetDataSetVersionId == nextDataSetVersionId,
+                cancellationToken);
+
+        AutoMapLocations(mappings.LocationMappingPlan);
+        AutoMapFilters(mappings.FilterMappingPlan);
+
+        mappings.LocationMappingsComplete = !mappings
+            .LocationMappingPlan
+            .Levels
+            .Any(level => level
+                .Mappings
+                .Any(optionMapping => IncompleteMappingTypes.Contains(optionMapping.Type)));
+
+        mappings.FilterMappingsComplete = !mappings
+            .FilterMappingPlan
+            .Mappings
+            .Any(filterMapping =>
+                filterMapping.Type == MappingType.None
+                || filterMapping
+                    .OptionMappings
+                    .Any(optionMapping => IncompleteMappingTypes.Contains(optionMapping.Type)));
+
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+    }
+
+    private static void AutoMapLocations(LocationMappingPlan locationPlan)
+    {
+        locationPlan
+            .Levels
+            .ForEach(level => level
+                .Mappings
+                .ForEach(locationMapping => AutoMapElement(
+                    mapping: locationMapping,
+                    candidates: level.Candidates)));
+    }
+
+    private static void AutoMapFilters(FilterMappingPlan filtersPlan)
+    {
+        filtersPlan
+            .Mappings
+            .ForEach(filterMapping => AutoMapParentAndOptions(
+                parentMapping: filterMapping,
+                parentCandidates: filtersPlan.Candidates,
+                candidateOptionsSupplier: autoMappedCandidate => autoMappedCandidate.Options));
+    }
+
+    private static TCandidate? AutoMapElement<TMappableElement, TCandidate>(
+        Mapping<TMappableElement> mapping,
+        List<TCandidate> candidates)
+        where TMappableElement : MappableElement
+        where TCandidate : MappableElement
+    {
+        if (candidates.Count == 0)
+        {
+            mapping.Type = MappingType.AutoNone;
+            return null;
+        }
+
+        var matchingCandidate = candidates
+            .SingleOrDefault(candidate =>
+                candidate.Key == mapping.Source.Key);
+
+        if (matchingCandidate is not null)
+        {
+            mapping.CandidateKey = matchingCandidate.Key;
```

Comment:
Yeah it could go either way tbh.  You're right that we always set it to the same key as the source during automapping, but we allow the user to update it as well.  We might have brainier strategies that look at other metadata in the future and don't rely solely on identical keys to perform an automapping.

I could go either way namewise, but probably err on CandidateKey just so it's clearer what we're mapping to if that's OK with you?

Reply:
Yep, all good with me! :) I could also go either way